lib/state/state.js

```diff
@@ -517,36 +521,72 @@ State.prototype.respond = function (response, isJson) {
 	this.response = isJson ? response : JSON.stringify(response);
 };
 
+function destroy(error, state, response, cb) {
+	// cleanup
+	state.destroy();
+
+	// return the response
+	if (cb) {
+		cb(null, response);
+	}
+}
+
+function respond(error, state, cb) {
+	const myEvents = parseEvents(error, state.myEvents);
+	state.myEvents = undefined;
+
+	destroy(error, state, {
+		response: state.response,
+		myEvents: myEvents.length ? myEvents : undefined
+	}, cb);
+}
+
 
-function closeWithError(state, cb) {
-	cb(null, {
-		errorCode: state.errorCode,
-		myEvents: state.myEvents.length ? state.myEvents : undefined
+function close(error, state, cb) {
+	state.emitEvents(error, function () {
+		respond(error, state, cb);
 	});
 }
 
 
 function closeWithSuccess(state, cb) {
-	state.archivist.distribute(function (error) {
-		// archivist errors can only come from the operations before writing anything to a vault,
-		// (beforeDistribute hooks, operation validation, etc), so it is still safe to bail out and
-		// report an error.
+	// archivist errors can only come from the operations before writing anything to a vault,
+	// (beforeDistribute hooks, operation validation, etc), so it is still safe to bail out and
+	// report an error.
+	state.archivist.distribute((error) => close(error, state, cb));
+}
+
+/**
+ * Distribute all events currently stacked on
+ * this state object.
+ *
+ * If you wish to also distribute archivist mutations stacked
+ * on this state, please have a look at the `distribute` method
+ * instead.
+ *
+ * @param {Function} cb callback function
+ */
+State.prototype.distributeEvents = function (cb) {
+	Array.prototype.push.call(this.events, this.myEvents);
```

Comment:
Unrelated, but don't you mean to do `apply` instead of `call`? You are currently adding `this.myEvents` as a single element to `this.events`.